### ISSUE TITLE
Consistency between the data/property divisions and definitions of structured equivalence in Algebra

### ISSUE THREAD

**Issue by ecavallo** [agda/cubical#546]
In `Algebra` many of the definitions are divided into "structure" and "property" parts.
```
  record IsRing {R : Type ℓ}
              (0r 1r : R) (_+_ _·_ : R → R → R) (-_ : R → R) : Type ℓ where
  field
    +IsAbGroup : IsAbGroup 0r _+_ -_
    ·IsMonoid  : IsMonoid 1r _·_
    dist        : (x y z : R) → (x · (y + z) ≡ (x · y) + (x · z))
                              × ((x + y) · z ≡ (x · z) + (y · z))
```
```
record RingStr (A : Type ℓ) : Type (ℓ-suc ℓ) where

  constructor ringstr

  field
    0r      : A
    1r      : A
    _+_     : A → A → A
    _·_     : A → A → A
    -_      : A → A
    isRing  : IsRing 0r 1r _+_ _·_ -_
```
On top of these we have a definition of structured equivalence:
```
record RingEquiv (R : Ring {ℓ}) (S : Ring {ℓ'}) (e : ⟨ R ⟩ ≃ ⟨ S ⟩) : Type (ℓ-max ℓ ℓ') where

  constructor ringequiv

  private
    module R = RingStr (snd R)
    module S = RingStr (snd S)

  field
    pres1 : equivFun e R.1r ≡ S.1r
    isHom+ : (x y : ⟨ R ⟩) → equivFun e (x R.+ y) ≡ equivFun e x S.+ equivFun e y
    isHom· : (x y : ⟨ R ⟩) → equivFun e (x R.· y) ≡ equivFun e x S.· equivFun e y
```
The latter also implicitly makes a distinction between structure and property, as we check that each piece of structure is preserved but don't need to check this for properties.

However you can see in this example that the two don't line up: `0r` is put with the structure in the definition of ring structure, but left out (treated as a property) in the definition of the structured equivalence. This discrepancy gets in the way of automating the proof that `RingEquiv`s correspond to paths in `RingStr` and so on. The same problem exists with `-_` here. I'm not sure if there are other mismatched examples in the library, though everything based on rings has the `0r` and `-_` problem.

Judging by #310, I guess the preferred solution is to require proofs of `0r` and `-_` preservation in the equivalence and then add helper functions for generating these, so that we leave open the option to handcraft a proof if desired?

**Comment by ecavallo** [comment]
Requesting @felixwellen's opinion in particular

**Comment by felixwellen** [comment]
Yes, I agree. It should be no problem to write constructors for isomorphisms. I can help with the refactoring (might take some time though). 
There are two related issues, we might want to fix while we are at it:
* Definitions of morphisms and isomorphism are separated for some structures. I guess it would be nicer, to define isomorphisms as morphisms that are equivalences. (That is #406)
* Algebra and CommAlgebras don't follow the pattern of all the other algebraic structures. That should be fixed, but it might be actual work. And it should have influence on the 'right' definition for morphisms.

Maybe you can start a PR, where the morphism definitions are as they should be for rings and I continue it?

**Comment by ecavallo** [comment]
>  Definitions of morphisms and isomorphism are separated for some structures. I guess it would be nicer, to define isomorphisms as morphisms that are equivalences. (That is #406)

Once we finish getting the current Structure machinery replaced with DUARels, I do want to set up something analogous for morphisms, but that may not happen for a while.

> Maybe you can start a PR, where the morphism definitions are as they should be for rings and I continue it?

Sounds good, I'll have a chance in a few days.

@MatthewDaggitt's suggestion sounds like a good idea to me, though it's mostly orthogonal to the issue I'm raising (which should be fixed whether we change formats or not). We're deviating from the standard library by doing some flattening: for us, a structure is defined as a record one of whose fields is a collection of properties, rather than a collection of structure paired with a collection of properties. The pair version definitely looks better from a usability perspective. I guess we were hoping to avoid an extra level of nesting for (nebulous?) efficiency reasons, but I'm suspicious that it's worth it.

**Comment by felixwellen** [comment]
> Sounds good, I'll have a chance in a few days.

Great! Next week, I should have a little bit of time to do something...

**Comment by ecavallo** [comment]
Resolved by #557.
